pkg/controller/pv-monitor-controller/node_watcher.go

```diff
@@ -0,0 +1,455 @@
+			continue
+		}
+
+		if pv.Status.Phase != v1.VolumeBound {
+			continue
+		}
+
+		pods := watcher.pvcToPodsCache.GetPodsByPVC(pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)
+		if pods == nil || len(pods) == 0 {
+			continue
+		}
+
+		podsOnThatNode := make([]*v1.Pod, 0)
+		for _, pod := range pods {
+			if pod.Spec.NodeName == node.Name {
+				podsOnThatNode = append(podsOnThatNode, pod)
+			}
+		}
+		if len(podsOnThatNode) == 0 {
+			continue
+		}
+
+		pvc, err := watcher.pvcLister.PersistentVolumeClaims(pv.Spec.ClaimRef.Namespace).Get(pv.Spec.ClaimRef.Name)
+		if err != nil {
+			klog.Errorf("get PVC[%s] from PVC lister error: %+v", pv.Spec.ClaimRef.Namespace+"/"+pv.Spec.ClaimRef.Name, err)
+			return err
+		}
+
+		pvcClone := pvc.DeepCopy()
+
+		message := "Node: " + node.Name + " recovered"
+		watcher.recorder.Event(pvcClone, v1.EventTypeWarning, "NodeRecovered", message)
+
+		// TODO: handle PVC conditions for the node recovery?
+		/*nodeFailureExist := false
+		for _, pvcCondition := range pvcClone.Status.Conditions {
+			if pvcCondition.Type == util.NodeFailed + "d" {
+				pvcCondition.Message = "Node:"
+				pvcCondition.Type = util.NodeRecovered
+				nodeFailureExist = true
+				break
+			}
+		}
+
+		if !nodeFailureExist {
+			continue
+		}
+
+		_, err = watcher.client.CoreV1().PersistentVolumeClaims(pvcClone.Namespace).UpdateStatus(pvcClone)
+		if err != nil {
+			klog.Errorf("update PVC[%s] error: %+v", pvcClone.Namespace + "/" + pvcClone.Name, err)
+			return err
+		}*/
+
+	}
+	return nil
+}
+
+func (watcher *NodeWatcher) markPVCsAndPodsOnUnhealthyNode(node *v1.Node) error {
+	pvs, err := watcher.volumeLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Warningf("cannot list pvs: %s", err)
+		return err
+	}
+
+	for _, pv := range pvs {
+		if pv.Spec.CSI == nil || pv.Spec.CSI.Driver != watcher.driverName {
+			continue
+		}
+
+		if pv.Status.Phase != v1.VolumeBound {
+			continue
+		}
+
+		pods := watcher.pvcToPodsCache.GetPodsByPVC(pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)
+		if pods == nil || len(pods) == 0 {
+			continue
+		}
+
+		podsOnThatNode := make([]*v1.Pod, 0)
+		for _, pod := range pods {
+			if pod.Spec.NodeName == node.Name {
+				podsOnThatNode = append(podsOnThatNode, pod)
+			}
+		}
+		if len(podsOnThatNode) == 0 {
+			continue
+		}
+
+		pvc, err := watcher.pvcLister.PersistentVolumeClaims(pv.Spec.ClaimRef.Namespace).Get(pv.Spec.ClaimRef.Name)
+		if err != nil {
+			klog.Errorf("get PVC[%s] from PVC lister error: %+v", pv.Spec.ClaimRef.Namespace+"/"+pv.Spec.ClaimRef.Name, err)
+			return err
+		}
+
+		pvcClone := pvc.DeepCopy()
+
+		message := "Pods: [ "
+		for _, pod := range podsOnThatNode {
+			message = message + pod.Name + " "
+		}
+		message += "]" + " consuming PVC: " + pvcClone.Name + " in namespace: " + pvcClone.Namespace + " are now on a failed node: " + node.Name
+
+		watcher.recorder.Event(pvcClone, v1.EventTypeWarning, "NodeFailed", message)
```

Comment:
In the CSI spec, we have stated that the messages are coming from GetVolume or ListVolumes on the controller side.  It maybe confusing to add this here.

We need to discuss where to log this event.

Reply:
Discussed this at the CSI Implementation meeting today.  It's fine to add this annotation on PVC, but we should provide a flag to make this node_watcher configurable because not every driver needs it.